Core contract l3 interface

aniketpr01 opened the pull request:
Setup Core Contract Interface for L2 <-> L3 Interaction

Summary
This PR introduces a core contract interfaces within the l3 folder to facilitate interactions between an appchain and StarkNet. 

Major Changes
1. Core contract Interfaces: Added interfaces for Operator, Messaging and Update State.
2. StarknetClient Struct: Added a StarknetClient struct that encapsulates the client and URL for interacting with StarkNet.
3. Client Method: Added a client method to the StarknetClient struct to return a reference to the LocalWalletSignerMiddleware.
4. Attach Method: Added an attach method to the StarknetClient struct to initialize and return a new instance of StarknetClient. This method takes in parameters such as rpc_endpoint, chain_id, priv_key, and account_addr to set up the client.
6. Restructured existing ethereum contract interfaces inside L2 folder, and all the changes as mentioned above in L3 folder.

aniketpr01 committed:
added submodule for piltover

aniketpr01 committed:
added draft implementation of register_operator

aniketpr01 committed:
added draft implementation of register_operator

aniketpr01 committed:
raw incomplete implementation, pending TODOs

aniketpr01 committed:
updated initial working implementation draft for starknet client

aniketpr01 committed:
starknetclient created, interfaces pending

aniketpr01 committed:
pushing a structured interface implementation for operator

aniketpr01 committed:
adding cargo.toml back

aniketpr01 committed:
implemented register_operator method

apoorvsadana commented on .gitignore~:
```diff
@@ -16,4 +16,7 @@ docs/
+.idea
+.vscode
+.python-version
+Cargo.lock
```
Cargo.lock must not be ingored

apoorvsadana replied:
is the file rename a bug or does it solve something

aniketpr01 replied:
oh yes, my bad, this was not intentional, i will remove from from gitignore and add back

apoorvsadana commented on crates/l3/appchain-core-contract-client/artifacts/piltover.starknet_artifacts.json:
do we need the artifacts file?

apoorvsadana commented on crates/l3/appchain-core-contract-client/artifacts/piltover_appchain.compiled_contract_class.json:
do we need to add all files in the repo directly? can we not add the git submodule and add the building part in the makefile? i know we didn't do that for the bridging contracts. over there, we had some issues in adding the git submodule, but I guess for this one it should be possilble. not a priority for now, we can do the interfaces first but keeping the comment

aniketpr01 replied:
sure, will take this next

apoorvsadana replied:
this is still pending i guess?

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/clients/mod.rs:
```diff
@@ -0,0 +1,3 @@
+mod client;
+
+pub use client::*;
```
we can just do `pub mod client`?

aniketpr01 replied:
yes yes!

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/mod.rs:
same as above

aniketpr01 replied:
sure!

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,65 @@
+}
+
+#[async_trait]
+pub trait OperatorTrait {
```
hey, now that I see it, I think we can skip having traits at all. basically, in ethers rs , we needed traits to abstract the functioning of ethers rs because the abigen! implemented some methods but they could not be called directly. we had to abstract them using our traits. however, here, since we don't have abigen!, we can do the abstraction in our methods itself.

sorry, i should've caught this earlier, but let's remove the traits for now to save time later on. you can just remove the trait and rest of the code should remain the same I believe.

aniketpr01 replied:
makes sense, will remove it!

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/lib.rs:
```diff
@@ -0,0 +1,28 @@
+const STARKNET_CORE_CONTRACT_SIERRA: &str =
+    include_str!("../artifacts/piltover_appchain.contract_class.json");
+
+pub async fn deploy_starknet_core_contract(client: LocalWalletSignerMiddleware) {
```
can you add a TODO here to check if we need to do this behind a proxy?

aniketpr01 replied:
will do!

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/main.rs:
were you using this file for testing? we don't need a main.rs in this repo, it's only going to be used as a library

aniketpr01 replied:
oh yes, it was for testing only!

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,35 @@
+    SingleOwnerAccount<JsonRpcClient<HttpTransport>, LocalWallet>;
+pub const NO_CONSTRUCTOR_ARG: () = ();
+
+pub trait StarknetContractClient {
```
there are two `StarknetContractClient` currently, one here and the other in the utils folder

aniketpr01 replied:
my bad, shall i rename this to StarknetContractClientL3?

apoorvsadana replied:
isn't it the same thing, do we need both?

aniketpr01 replied:
yes, one is using middleware from ethers library and other from starknetone

apoorvsadana replied:
ahh I see now, is the `StarknetContractClient` in common (for l3s) being used anywhere? I don't see it

apoorvsadana replied:
maybe we want to implement it for `StarknetCoreContractClient`. we can call the one in utils as `EthereumContractClient`

aniketpr01 replied:
sure, have left a todo, will change when doing restructuring of ethereum interfaces

apoorvsadana replied:
let's restructure in this PR itself? ig we just need to move to a new folder right, or am I missing something

apoorvsadana replied:
next step would be PRs on dependent repos which is only Madara for now I guess

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+    let contract_factory = ContractFactory::new(class_hash, client);
+
+    let deploy_tx = &contract_factory.deploy(
+        vec![FieldElement::from_hex_be("0x100000").unwrap()],
```
should be constructor_args?

aniketpr01 replied:
will fix this!

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+            provider,
+            signer,
+            account_address,
+            FieldElement::from_hex_be("0x4b4154414e41").unwrap(),
```
should be an argument

aniketpr01 replied:
for which section/s?

apoorvsadana replied:
the chain id should be taken as an argument and not hardcoded

apoorvsadana replied:
it's still hardcoded, it should be an argument to the function

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,114 @@
+    let class_hash = sierra.class_hash().unwrap();
+    let compiled_class_hash = compiled_class_hash;
+
+    println!("Contract declared at address: {:?}", result);
+    println!("Compiled Class hash : {:?}", compiled_class_hash);
```
we can make these debug logs

aniketpr01 replied:
will do!

aniketpr01 committed:
implemented Operator

aniketpr01 commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,43 @@
+    //     .await;
+    //     // let nonce = execution.estimate_fee(); // how to get the nonce from the execution
+
+
```
@apoorvsadana 

As per interface implemenation of messaging contract, the return value is a tuple of two FieldElement
It Returns the message hash and the updated nonce of the message.

Exisiting implementation of executing contract in starknet-rs doesn't return anything
so, how can we read/return the message hash and the updated nonce of the message?

There is a method which can calculate hash as defined here but not sure how to use that:
https://github.com/keep-starknet-strange/piltover/blob/ff8853fde770fac31ea071892242e12cd386a3f1/src/messaging/hash.cairo#L21

aniketpr01 committed:
implemented messaging and core_contract interfaces

aniketpr01 committed:
cargo fmt

aniketpr01 committed:
removed main.rs from appchain structure

aniketpr01 committed:
restructured and optimized

aniketpr01 committed:
reverted gitignore

aniketpr01 committed:
reverted gitignore

apoorvsadana commented on .python-version:
```diff
@@ -0,0 +1 @@
+starknet_env
```
do we need this?

aniketpr01 replied:
removed

apoorvsadana commented on .vscode/settings.json:
should we add it to git ignore?

aniketpr01 replied:
removed for now since this can be individual settings for vscode

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    pub async fn is_operator(&self) -> Result<bool, StarknetError> {
+        let result = call_contract(self.client.clone(), self.address, "is_operator").await;
```
can we pass client ref instead of cloning?

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,115 @@
+                    Err(StarknetError::ContractError)
+                }
+            }
+            None => Err(StarknetError::ContractError),
```
we should return a Result from call_contract instead of an Option and throw the exact error here?

aniketpr01 replied:
I think we are returning error from call contract but in case there is any malformed request or any other issue, ContractError is raised, should we ignore this error and relying on error from call_contract would be sufficient?

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,53 @@
+) -> Execution<'a, LocalWalletSignerMiddleware> {
+    let call = Call {
+        to: address,
+        selector: get_selector_from_name(method.into()).unwrap(),
```
we should throw the error instead on unwrap

aniketpr01 replied:
done

apoorvsadana replied:
i think this is still pending?

apoorvsadana commented on crates/l3/common/src/errors.rs:
these errors aren't being used anywhere? maybe you want throw these errors from contract call and invokes

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,53 @@
+    SingleOwnerAccount<JsonRpcClient<HttpTransport>, LocalWallet>;
+pub const NO_CONSTRUCTOR_ARG: Vec<FieldElement> = Vec::new();
+
+pub trait StarknetContractClientL3 {
```
let's just call it `StarknetContractClient`

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,53 @@
+        .await
+    {
+        Ok(result) => Some(result),
+        Err(_) => None,
```
we should throw the error

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+    Hex(#[from] hex::FromHexError),
+    #[error("Failed to parse URL")]
+    UrlParser,
+    //#[error(transparent)]
+    //EthersContract(#[from] ContractError<LocalWalletSignerMiddleware>),
+    //#[error(transparent)]
+    //EthersProvider(#[from] ProviderError),
+    //#[error("Invalid contract build artifacts: missing field `{0}`")]
+    //ContractBuildArtifacts(&'static str),
```
can remove the commented ones and the ones not being used as well

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+}
+
+pub async fn deploy_contract(
```
should we move this to common as well?

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+        let url = Url::parse(rpc_endpoint.unwrap().as_str()).expect("Invalid JSONRPC Url");
+        let provider = JsonRpcClient::new(HttpTransport::new(url.clone()));
+        let signer = LocalWallet::from(SigningKey::from_secret_scalar(
+            FieldElement::from_hex_be(priv_key.unwrap().as_str()).unwrap(),
```
let's use expect instead on unwrap

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+        let signer = LocalWallet::from(SigningKey::from_secret_scalar(
+            FieldElement::from_hex_be(priv_key.unwrap().as_str()).unwrap(),
+        ));
+        let account_address = FieldElement::from_hex_be(account_addr.unwrap().as_str())
```
expect instead of unwrap

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+
+pub async fn deploy_contract(
+    client: Arc<LocalWalletSignerMiddleware>,
+    contract_build_sierra: &str,
```
let's rename to `path_to_sierra` and `path_to_casm`? i guess it will be easier to understand that way

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+            compiled_class_hash,
+        )
+        .max_fee(FieldElement::from_hex_be("0x100000").unwrap());
+    let _result = declare_tx.send().await.unwrap();
```
let's remove unwraps and throw the error

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+    constructor_args: Vec<FieldElement>,
+) -> FieldElement {
+    let sierra: SierraClass = serde_json::from_reader(
+        std::fs::File::open(env!("CARGO_MANIFEST_DIR").to_owned() + "/" + contract_build_sierra)
```
i think we can directly take as input the entire path. we don't know how other programs want to use zaun, they might have the files stored elsewhere

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+            sierra.clone().flatten().unwrap().into(),
+            compiled_class_hash,
+        )
+        .max_fee(FieldElement::from_hex_be("0x100000").unwrap());
```
do we need to specify max_fee? i think it will automatically be estimated

aniketpr01 replied:
i think so, removing for now

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+        .max_fee(FieldElement::from_hex_be("0x100000").unwrap());
+    let _result = declare_tx.send().await.unwrap();
+    let class_hash = sierra.class_hash().unwrap();
+    let _compiled_class_hash = compiled_class_hash;
```
we can remove the `_` variables?

aniketpr01 replied:
done

aniketpr01 committed:
optimized and fixed code issues as per git review

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+    client.as_ref().execute(vec![call])
+}
+
+// pub async fn call_contract(
+//     client: Arc<LocalWalletSignerMiddleware>,
+//     address: FieldElement,
+//     method: &str,
+// ) -> Option<Vec<FieldElement>> {
+//     let function_call = FunctionCall {
+//         contract_address: address,
+//         entry_point_selector: get_selector_from_name(method.into()).unwrap(),
+//         calldata: vec![],
+//     };
+//     let provider = client.provider();
+//     match provider
+//         .call(function_call, BlockId::Tag(BlockTag::Latest))
+//         .await
+//     {
+//         Ok(result) => Some(result),
+//         Err(_) => None,
+//     }
+// }
```
let's remove this

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+    provider
+        .call(function_call, BlockId::Tag(BlockTag::Latest))
+        .await
+        .map_err(|e| Error::CustomError(format!("Failed to call contract: {}", e)))
```
the `Error` struct already has a `ProviderError` right which can be used here?

aniketpr01 replied:
handled

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+        .call(function_call, BlockId::Tag(BlockTag::Latest))
+        .await
+        .map_err(|e| Error::CustomError(format!("Failed to call contract: {}", e)))
+        .map(|result| result)
```
we can remove this map?

aniketpr01 replied:
done, my bad, redundant one!

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+) -> Result<Vec<FieldElement>, Error> {
+    let function_call = FunctionCall {
+        contract_address: address,
+        entry_point_selector: get_selector_from_name(method.into()).unwrap(),
```
let's throw an error instead of unwrap

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+    let declare_tx = client
+        .declare(
+            sierra.clone().flatten()
+                .map_err(|e| Error::CustomError(format!("Failed to flatten Sierra class: {}", e)))?.into(),
```
instead of using Custom Errors everywhere, you should check the color_eyre package, it should be the correct use case over here

aniketpr01 replied:
@apoorvsadana 
How should this be implemented, since when these errors then will be required to propagated to change across all the methods where it is called and will be called from other repo as well. 
Or should we create a enum Error with all specific errors and use that?

apoorvsadana replied:
So I was thinking of using eyre errors in those places where we were using CustomErrors, I guess we replaced all of them? Anyways, I think this works as well, later when we need to handle any error specifically, we can see what to do.

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    pub async fn is_operator(&self) -> Result<bool, Error> {
+        let result = call_contract(&self.client, self.address, "is_operator").await;
```
we can just do await`?` and it will throw the error automatically. also do run cargo fmt and cargo clippy on the code once

aniketpr01 replied:
kindly review this change

aniketpr01 committed:
fixed return errors and handled unwraps

aniketpr01 committed:
made changes to handle error using color-eyre package

aniketpr01 committed:
changed error handling for call_contract methods

aniketpr01 committed:
removed unwanted files and reverted .gitignore changes

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,52 @@
+            .map_err(|_| Error::Hex(hex::FromHexError::InvalidStringLength))?;
+        let signer = LocalWallet::from(SigningKey::from_secret_scalar(signer_key));
+
+        let account_address = FieldElement::from_hex_be(&account_addr.ok_or(Error::UrlParser)?)
```
not a UrlParser error

apoorvsadana replied:
I think we should pass the original error object as an argument to our error types otherwise the context is lost. So UrlParser takes an argument of ParseError for example

apoorvsadana replied:
this is left

aniketpr01 committed:
updated makefile to compile piltover contracts, handled error formatting

apoorvsadana commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,90 @@
+        to_address: FieldElement,
+        selector: FieldElement,
+        payload: Vec<FieldElement>,
+        // ) -> Result<Option<Execution<LocalWalletSignerMiddleware>>, Error> {
```
let's remove this comment, and any other code comments we've in code

aniketpr01 replied:
done

apoorvsadana commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,50 @@
+
+impl StarknetClient {
+    pub fn attach(
+        rpc_endpoint: Option<String>,
+        chain_id: Option<String>,
+        priv_key: Option<String>,
+        account_addr: Option<String>,
```
why take optional if we're going to throw an error if they're not present

aniketpr01 replied:
oh yeah, that will be better, thank you!

aniketpr01 committed:
moved ethereum interfaces to l2 and fixed error handling for starknetclient

aniketpr01 committed:
refactored starknet interface attach method

aniketpr01 committed:
Refactor contract function calls and reordering of imports.
Simplified the way contract functions
are called by passing the calldata
directly to the invoke_contract
function, which improves code
readability and maintainability.

aniketpr01 committed:
refactored and removed unnecessary files

aniketpr01 committed:
Add support for lifetime parameters and added test

aniketpr01 committed:
updated by cargo fmt

apoorvsadana commented on crates/l3/appchain-core-contract-client/tests/test_client.rs:
```diff
@@ -0,0 +1,37 @@
+    let starknet_client = StarknetClient::attach(rpc_endpoint, chain_id, priv_key, account_addr)?;
+
+    // Create a StarknetCoreContractClient
+    let address = FieldElement::from_hex_be("0x07100a5f88487652c9efaa81d9037927f67000512ff25e00dded29e5af58ae58")?;
```
is this supposed to be predeployed on devnet? if yes, I think it's better to deploy it as a part of the test no? otherwise how do we test it in CI

aniketpr01 replied:
I will create another issue in which i will write tests in a sandboxed devnet environment.

aniketpr01 committed:
reverted functions to accept original function signatures

aniketpr01 committed:
removed tests for now which will be added in another issue/PR

tdelabro commented on crates/l3/common/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "common"
```
terrible package name

tdelabro replied:
imagine the package is published on crate.io. This would be it's name.

We can find a more descriptive name.

tdelabro replied:
moreover it looks like it is mostly used in `appchain-core-contract-client`. So maybe it should just live there as `error` and `utils` modules

aniketpr01 replied:
noted.
I can rename it to appchain-utils but i believe this will be required as methods invoke_contract and call_contract will be used for another module like verifier contract to be added next hence kept it as a common/utils package.

tdelabro commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+}
+
+pub async fn invoke_contract(
+    client: &Arc<&LocalWalletSignerMiddleware>,
```
It looks like you are only using `client` to do `client.as_ref()` so can you change the signature so that the type of `client` is whatever `client.as_ref()` returns?

aniketpr01 replied:
updated to return &LocalWalletSignerMiddleware as client.as_ref() returns a reference to it.

tdelabro commented on crates/l3/common/src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+        entry_point_selector,
+        calldata,
+    };
+    let provider = client.provider();
+    provider
+        .call(function_call, BlockId::Tag(BlockTag::Latest))
```
Same. It looks like you are using client only to get `provider` out of it.
Can you change the signature so that it takes the provider directly as argument?

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/lib.rs:
```diff
@@ -0,0 +1,28 @@
+        path_to_casm,
+        NO_CONSTRUCTOR_ARG,
+    )
+    .await?;
```
space before return for readablility

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/clients/mod.rs:
```diff
@@ -0,0 +1 @@
+pub mod client;
```
just put the content of `client.rs` here. No need for two files

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,88 @@
+use std::sync::Arc;
+
+pub struct Messaging<'a> {
+    client: Arc<&'a LocalWalletSignerMiddleware>,
```
It is not a `client` per say. Just the signer middleware.

It seems to me that the name `client` has been used abusively all over this crate. Don't you think? 
Client is `StarknetCoreContractClient` defined in `clients/client`

aniketpr01 replied:
yes agreed 🫡 have refactored to be signer.

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/core_contract.rs:
```diff
@@ -0,0 +1,29 @@
+use std::sync::Arc;
+
+pub struct CoreContract<'a> {
+    client: Arc<&'a LocalWalletSignerMiddleware>,
```
Can't you just store `&'a LocalWalletSignerMiddleware` here? Rather than the arc?

aniketpr01 replied:
makes sense, will not use the Arc for now and simply passing as reference should be fine.

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/core_contract.rs:
```diff
@@ -0,0 +1,29 @@
+        calldata.extend(program_output);
+        calldata.push(onchain_data_hash);
+        calldata.push(onchain_data_size);
+        invoke_contract(&self.client, self.address, "update_state", calldata).await
```
space before return line

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,88 @@
+        let mut calldata = Vec::new();
+        calldata.push(to_address);
+        calldata.push(selector);
+        calldata.extend(payload);
```
space before return

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,88 @@
+        let mut calldata = Vec::new();
+        calldata.push(from_address);
+        calldata.extend(payload);
+        invoke_contract(
```
space before return

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,88 @@
+        calldata.push(selector);
+        calldata.push(nonce);
+        calldata.extend(payload);
+        invoke_contract(
```
space before return

aniketpr01 replied:
done

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,88 @@
+        calldata.push(selector);
+        calldata.push(nonce);
+        calldata.extend(payload);
+        invoke_contract(&self.client, self.address, "cancel_message", calldata).await
```
space before return

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+            "is_operator",
+            vec![operator.into()],
+        )
+        .await?;
```
space before retrun

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+use std::sync::Arc;
+
+pub struct Operator<'a> {
+    client: Arc<&'a LocalWalletSignerMiddleware>,
```
same. It is not the client here. And can you get rid of the `Arc` or is it required for something?

aniketpr01 replied:
yeah so the structure followed was based on the existing implementation of ethereum contract clients, where Arc is passed for thread safety i believe, as the client passed from other repo(in this case madara/orchestrator) was with Arc::new(). should we keep this or can be optimised?

tdelabro replied:
But its content is an immutable reference. So it doesn't add any safety here. Does it?

aniketpr01 replied:
makes sense, will not use the Arc for now and simply passing as reference should be fine.

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+
+    pub async fn get_program_info(&self) -> Result<(FieldElement, FieldElement)> {
+        let values = call_contract(&self.client, self.address, "get_program_info", vec![]).await?;
+        if values.len() == 2 {
```
space before return

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+    pub async fn get_facts_registry(&self) -> Result<FieldElement> {
+        let values =
+            call_contract(&self.client, self.address, "get_facts_registry", vec![]).await?;
+        if let Some(value) = values.first() {
+            Ok(value.clone())
+        } else {
+            Err(eyre!("Contract error: expected at least one return value"))
+        }
```
`values.first().or_or_else(|| eyre!("Contract error: expected at least one return value"))`

tdelabro replied:
more rustacean syntax

aniketpr01 replied:
done

tdelabro commented:
also @aniketpr01 can you give a summary of what is going on in this PR from a high level pov? In the PR description

aniketpr01 committed:
The module  has been renamed to  to better reflect its utilization. Consequently, dependencies and import statements referring to  have been updated. Additionally,  has been moved out of its  subdirectory and directly implemented under the  module.

aniketpr01 commented:
> also @aniketpr01 can you give a summary of what is going on in this PR from a high level pov? In the PR description

@tdelabro Requesting review again.

tdelabro commented:
@aniketpr01 I answer a few questions. And there is still some comments you didn't address from the previous one

aniketpr01 committed:
refactored client word usage to signer and removed Arc

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/core_contract.rs:
```diff
@@ -0,0 +1,29 @@
+        onchain_data_hash: FieldElement,
+        onchain_data_size: FieldElement,
+    ) -> Result<InvokeTransactionResult> {
+        let mut calldata = Vec::new();
```
You can use `Vec::with_capacity(program_output.len() + 2)` instead

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,91 @@
+        selector: FieldElement,
+        payload: Vec<FieldElement>,
+    ) -> Result<InvokeTransactionResult> {
+        let mut calldata = Vec::new();
```
with capacity

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,91 @@
+        from_address: FieldElement,
+        payload: Vec<FieldElement>,
+    ) -> Result<InvokeTransactionResult> {
+        let mut calldata = Vec::new();
```
with capacity

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,91 @@
+        nonce: FieldElement,
+        payload: Vec<FieldElement>,
+    ) -> Result<InvokeTransactionResult> {
+        let mut calldata = Vec::new();
```
with_capacity

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/messaging.rs:
```diff
@@ -0,0 +1,91 @@
+        nonce: FieldElement,
+        payload: Vec<FieldElement>,
+    ) -> Result<InvokeTransactionResult> {
+        let mut calldata = Vec::new();
```
with capacity

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+        new_operator: FieldElement,
+    ) -> Result<InvokeTransactionResult> {
+        invoke_contract(
+            &self.signer,
```
`self.signer` is already a ref. I'm not sure you need to reference it again given the signature expect a `signer: &LocalWalletSignerMiddleware`

tdelabro replied:
Check it and if so, fix it for all others occurrences too

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -0,0 +1,110 @@
+    method: &str,
+    calldata: Vec<FieldElement>,
+) -> Result<InvokeTransactionResult> {
+    let selector = get_selector_from_name(method.into())
```
unnecessary `into`?
`fn get_selector_from_name(func_name: &str)`

tdelabro replied:
If so, change it to all occurences

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -0,0 +1,110 @@
+
+pub const NO_CONSTRUCTOR_ARG: Vec<FieldElement> = Vec::new();
+
+pub const MAX_FEE: &str = "0x1000000000000";
```
try to create a const of type FieldElement directly, using `FieldElement::from_mont`.
Get the Montgomery repr by printing `FieldElement::into_mont` and then use it as input for `from_mont`

aniketpr01 replied:
thanks for this suggestion, understood the performance and potential errors related to string parsing during runtime if not done 🫡

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -0,0 +1,110 @@
+
+pub async fn deploy_contract<'a>(
+    signer: &'a LocalWalletSignerMiddleware,
+    path_to_sierra: &str,
+    path_to_casm: &str,
```
path args should probably of type `&Path` rather than `&str`

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -0,0 +1,110 @@
+    path_to_casm: &str,
+    constructor_args: Vec<FieldElement>,
+) -> Result<FieldElement> {
+    let sierra_file = std::fs::File::open(path_to_sierra)
+        .map_err(|e| eyre!("Failed to open Sierra file: {}", e))?;
+    let sierra: SierraClass = serde_json::from_reader(sierra_file)?;
```
You can improve readability this way:

```rust
let sierra: SierraClass = {
 let sierra_file = std::fs::File::open(path_to_sierra)
        .map_err(|e| eyre!("Failed to open Sierra file: {}", e))?;
    serde_json::from_reader(sierra_file)?
};
```

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -0,0 +1,110 @@
+
+    let contract_factory = ContractFactory::new(class_hash, signer);
+
+    let deploy_tx = &contract_factory.deploy(constructor_args, FieldElement::ZERO, true);
```
deprecated method

https://github.com/xJonathanLEI/starknet-rs/blob/56177d2aaf98cf8b95ee5fb8c77e0229ca088ee3/starknet-contract/src/factory.rs#L115C5-L123C6

tdelabro replied:
Also why do you borrow it?

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+            &self.signer,
+            self.address,
+            "register_operator",
+            vec![new_operator.into()],
```
no need to call `into`

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+            &self.signer,
+            self.address,
+            "unregister_operator",
+            vec![removed_operator.into()],
```
no need for `into`

tdelabro replied:
try run clippy

aniketpr01 replied:
gotcha, fixed 🙏

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+    pub async fn is_operator(&self, operator: FieldElement) -> Result<bool> {
+        let provider = self.provider();
+        let values =
+            call_contract(provider, self.address, "is_operator", vec![operator.into()]).await?;
```
no need for `operator.into`

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+            call_contract(provider, self.address, "is_operator", vec![operator.into()]).await?;
+
+        values.first()
+            .map(|value| value.to_string() != "0")
```
`value.to_string() != "0"` don't need to convert to string, you can compare to `FieldElement::ZERO`

Or even better, call `is_zero`

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -0,0 +1,103 @@
+            &self.signer,
+            self.address,
+            "set_facts_registry",
+            vec![facts_registry.into()],
```
into?

aniketpr01 replied:
fixed

tdelabro commented on crates/l3/starknet-instance/src/lib.rs:
```diff
@@ -0,0 +1,53 @@
+#[derive(Debug)]
+pub struct StarknetClient {
+    signer: LocalWalletSignerMiddleware,
+    url: Url,
```
I don't think this field is ever used. Is it?

aniketpr01 replied:
fixed

aniketpr01 committed:
refactored and fixed issues related to referencing, borrowing and conversions

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -1,6 +1,8 @@
-use color_eyre::{eyre::eyre, Result};
+// allow unused imports
+#![allow(unused_imports)]
```
Why ?

aniketpr01 replied:
it was done earlier as Provider was not being identified by client, but it's now working, removed!

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -18,7 +20,12 @@ pub type TransactionExecution<'a> = Execution<'a, RpcAccount<'a>>;
 pub const NO_CONSTRUCTOR_ARG: Vec<FieldElement> = Vec::new();
 
-pub const MAX_FEE: &str = "0x1000000000000";
+pub const MAX_FEE: FieldElement = FieldElement::from_mont([
```
Can you add a comment explaining which value it is?
```rust
/// 0x1000000000000
```

aniketpr01 replied:
done

tdelabro commented on crates/l3/appchain-utils/src/lib.rs:
```diff
@@ -31,17 +38,17 @@ pub async fn invoke_contract(
         calldata,
     };
-    let max_fee = FieldElement::from_hex_be(MAX_FEE).unwrap();
+    // let max_fee = FieldElement::from_hex_be(MAX_FEE).unwrap();
```
remove commented code

aniketpr01 replied:
done

aniketpr01 committed:
removed commented code, and unused import declaration

tdelabro commented on crates/l3/appchain-core-contract-client/src/interfaces/operator.rs:
```diff
@@ -1,14 +1,10 @@
 // allow unused imports
```
remove comment

aniketpr01 replied:
gotcha

aniketpr01 committed:
removed comment
